import { IOrder } from "../types/orders.types";

const API_BASE_URL = "http://localhost:5000/api";

export async function createOrder(data: IOrder) {
    try {
        const response = await fetch(`${API_BASE_URL}/shop/orders`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            credentials: "include",
            body: JSON.stringify(data),
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message);
        }

        return await response.json();
    } catch (error: any) {
        throw new Error(error?.message || "Помилка з'єднання із сервером");
    }
}

export async function getOrders(): Promise<IOrder[]> {
    try {
        const response = await fetch(`${API_BASE_URL}/admin/orders`, {
            credentials: "include",
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message);
        }

        return await response.json();
    } catch (error: any) {
        throw new Error(error?.message || "Помилка з'єднання із сервером");
    }
}

export async function getOrdersByUserId(): Promise<IOrder[]> {
    try {
        const response = await fetch(`${API_BASE_URL}/shop/orders/users`, {
            credentials: "include",
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message);
        }

        return await response.json();
    } catch (error: any) {
        throw new Error(error?.message || "Помилка з'єднання із сервером");
    }
}

export async function updateOrder(
    orderId: string | undefined,
    data: Partial<IOrder>
): Promise<IOrder[]> {
    try {
        const response = await fetch(`${API_BASE_URL}/admin/orders/${orderId}`, {
            method: "PATCH",
            headers: {
                "Content-Type": "application/json",
            },
            credentials: "include",
            body: JSON.stringify(data),
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message);
        }

        return await response.json();
    } catch (error: any) {
        throw new Error(error?.message || "Помилка з'єднання із сервером");
    }
}

export async function deleteOrder(orderId: string) {
    try {
        const response = await fetch(`${API_BASE_URL}/admin/orders/${orderId}`, {
            method: "DELETE",
            credentials: "include",
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message);
        }

        return await response.json();
    } catch (error: any) {
        throw new Error(error?.message || "Помилка з'єднання із сервером");
    }
}
